clang/lib/AST/Type.cpp

```diff
@@ -560,6 +560,27 @@ bool Type::isIntCapType() const {
   return false;
 }
 
+bool Type::canCarryProvenance(const ASTContext &C) const {
+  if (!isCHERICapabilityType(C)) {
+    // In pure-capability mode we know that only capabilities carry provenance
+    if (C.getTargetInfo().areAllPointersCapabilities())
+      return false;
+    // In certain cases in hybrid mode, pointer types can be implicitly
+    // converted to capabilities so even though they don't carry provenance,
+    // the resulting type might do.
+    // However, null pointers never carry provenance.
+    if (!isPointerType() || isNullPtrType())
+      return false;
+  }
```

Comment:
Do we really want to be continuing out of this `if` in hybrid mode? I'm concerned that the enum case might be wrong, and the comment at the end of the function talks about capabilities but not hybrid pointers.